source/common/stats/allocator_impl.cc

```diff
@@ -316,5 +337,41 @@ Counter* AllocatorImpl::makeCounterInternal(StatName name, StatName tag_extracte
   return new CounterImpl(name, *this, tag_extracted_name, stat_name_tags);
 }
 
+void AllocatorImpl::forEachSinkedCounter(std::function<void(std::size_t)> f_size,
+                                         std::function<void(Stats::Counter&)> f_stat) {
+  Thread::LockGuard lock(mutex_);
+  forEachSinkedStat(f_size, f_stat, counter_sink_filter_, counters_, sinked_counters_);
+}
+
+void AllocatorImpl::forEachSinkedGauge(std::function<void(std::size_t)> f_size,
+                                       std::function<void(Stats::Gauge&)> f_stat) {
+  Thread::LockGuard lock(mutex_);
+  forEachSinkedStat(f_size, f_stat, gauge_sink_filter_, gauges_, sinked_gauges_);
+}
+
+void AllocatorImpl::forEachSinkedTextReadout(std::function<void(std::size_t)> f_size,
+                                             std::function<void(Stats::TextReadout&)> f_stat) {
+  Thread::LockGuard lock(mutex_);
+  forEachSinkedStat(f_size, f_stat, text_readout_sink_filter_, text_readouts_,
+                    sinked_text_readouts_);
+}
+
+void AllocatorImpl::setCounterSinkFilter(
+    std::function<bool(const Stats::Counter&)> counter_sink_filter) {
+  ASSERT(counter_sink_filter_ == nullptr);
```

Comment:
I think you need to also iterate over any existing counters_ and put them into the sinked set too.

Should we allow counter_sink_filter to be set to nullptr here also, for symmetry? I don't think it's necessary but if you do that you should clear out the sinked set. Otherwise doc that's not allowed and assert.

Reply:
Will address in part 2.